sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -68,10 +76,23 @@ object NestedColumnAliasing {
    */
   def replaceChildrenWithAliases(
       plan: LogicalPlan,
+      nestedFieldToAlias: Map[ExtractValue, Alias],
       attrToAliases: Map[ExprId, Seq[Alias]]): LogicalPlan = {
     plan.withNewChildren(plan.children.map { plan =>
       Project(plan.output.flatMap(a => attrToAliases.getOrElse(a.exprId, Seq(a))), plan)
-    })
+    }).transformExpressions {
+      case f: ExtractValue if nestedFieldToAlias.contains(f) =>
+        nestedFieldToAlias(f).toAttribute
+    }
+  }
+
+  /**
+   * Returns true for those operators that we can prune nested column on it.
+   */
+  private def canPruneOn(plan: LogicalPlan) = plan match {
+    case _: Aggregate => true
+    case _: Expand => true
+    case _ => false
```

Comment:
Are the current entries all the case we can support for nested column pruning? How about `FlatMapGroupsInPandas`?

Reply:
I think FlatMapGroupsInPandas is also supported.

As you see, we do `transformExpressions` to replace `ExtractValue` in operator's expressions. So if the operator follows SparkSQL's operators, it should be fine.

This patch adds the change needed for general nested column pruning for the kind of operators which can prune nested column. Ideally we only need to add it in `canPruneOn`, and add test for it.

Currently I think nested column pruning test cases are all in Scala, no Python. I will think about how to add test.

Reply:
okay, thanks for the check.

Reply:
cc: @HyukjinKwon

Reply:
I think I was wrong. Re-checking `FlatMapGroupsInPandas`'s Python API, it looks like

```python
df.groupby("id").apply(normalize).show()
```

So basically the python udf takes no nested column selection but a full columns of DataFrame. It doesn't do nested column pruning.

Reply:
Sorry I just checked this. Yes, I think ^ is correct.